Manual/RecursiveDefs/WF/PreprocessExample.lean

```diff
@@ -123,19 +145,41 @@ theorem Pair.map_wfParam (f : α → β) (p : Pair α) :
 @[wf_preprocess]
 theorem Pair.map_unattach {P : α → Prop}
     (p : Pair (Subtype P)) (f : α → β) :
-    p.unattach.map f = p.map fun ⟨x, h⟩ => f (wfParam x) := by
+    p.unattach.map f =
+    p.map fun ⟨x, h⟩ =>
+      binderNameHint x f <|
+      f (wfParam x) := by
   cases p; simp [wfParam, Pair.unattach, Pair.map]
 ```
 
 Now the function body can be written without extra considerations, and the membership assumption is still available to the termination proof.
 
-```lean
+```lean (keep := false)
 def Tree.map (f : α → β) : Tree α → Tree β
   | leaf x => leaf (f x)
   | node p => node (p.map (fun t' => t'.map f))
 termination_by t => t
-decreasing_by have := Pair.size_lt_of_mem ‹_›; simp_all; omega
+decreasing_by
+  have := Pair.sizeOf_lt_of_mem ‹_›
+  simp_all
+  omega
 ```
 
+The proof can be made fully automatic by adding {name Pair.sizeOf_lt_of_mem}`sizeOf_lt_of_mem` to the {tactic}`decreasing_trivial` tactic, as is done for similar built-in theorems.
 
-:::
+```lean
+macro "sizeOf_pair_dec" : tactic =>
+  `(tactic| with_reducible
+    have := Pair.sizeOf_lt_of_mem ‹_›
+    omega
+    done)
```

Comment:
This is not really the general form for these tactics that one should use in a library (no simp after `have`, for example), but we hope that these ad-hoc tactics will be gone soon anyways, so maybe fine for now.